feat(helm): control scheduling of cilium pods

sypakine opened the pull request:
Fixes: #33462

This change allows users to control whether the anti-affinity rule is applied to cilium daemonset pods.

Pod anti-affinity rules may significantly reduce scheduling throughput.
See: https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity

```release-note
Control whether the anti-affinity rule is applied to cilium daemonset pods. Omitting the rule improves scheduling throughput for large clusters.
```

sypakine committed:
feat(helm): control scheduling of cilium pods

Fixes: #33462

This change allows users to control whether the anti-affinity rule is applied to cilium daemonset pods.

Pod anti-affinity rules may significantly reduce scheduling throughput. See: https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity

Signed-off-by: Mark St John <markstjohn@google.com>

gandro reviewed (approved):
As a non-default option, this seems fine to me.

Please note that the existing anti affinity rule exists to avoid race conditions where the Cilium DaewmonSet is uninstalled and reinstalled rather quickly. I'm not sure if the kube-scheduler takes care of this: https://github.com/cilium/cilium/issues/11828

sypakine committed:
docs(helm): add scaling section to performance tuning documentation

This section calls out the problem, how the scheduling mode can be used to alleviate it, as well as caveats.

Signed-off-by: Mark St John <markstjohn@google.com>

sypakine commented on Documentation/operations/performance/tuning.rst:
```diff
@@ -699,6 +699,29 @@ distributions offer kernel images with this option set or you can re-compile
 the Linux kernel. ``CONFIG_PREEMPT_NONE=y`` is the recommended setting for
 server workloads.
 
+Scaling
```
Sounds good, thanks.

sypakine commented on Documentation/operations/performance/tuning.rst:
```diff
@@ -699,6 +699,29 @@ distributions offer kernel images with this option set or you can re-compile
+=======
+
+By default, the `cilium daemonset <https://github.com/cilium/cilium/blob/main/install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml>`_
+is configured with an `inter-pod anti-affinity <https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity>`_
```
Ah, much better. Thanks for the reference.

sypakine commented on Documentation/operations/performance/tuning.rst:
```diff
@@ -699,6 +699,29 @@ distributions offer kernel images with this option set or you can re-compile
+Scaling
+=======
+
+By default, the `cilium daemonset <https://github.com/cilium/cilium/blob/main/install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml>`_
```
Sounds good, I'll remove it.
